### ISSUE TITLE
Playwright/Browsers hang when having more than 3 (sometimes 2) instances running at the same time when testing a Blazor Server App

### ISSUE THREAD

**Issue by dardanlibrary** [microsoft/playwright-dotnet#2203]
**Context:**
- Playwright Version: 1.23.0
- Operating System: Windows 11
- Node.js version: 16.13.0
- Browser: Chromium, Firefox

**Code Snippet**
```
using var playwright = await Playwright.CreateAsync();
await using var browser = await playwright.Firefox.LaunchAsync(
    new BrowserTypeLaunchOptions
    {
        Headless = false
    }
);

await Parallel.ForEachAsync(
    Enumerable.Range(1, options.Iterations),
    new ParallelOptions { MaxDegreeOfParallelism = options.Iterations },
    async (iteration, cancellationToken) =>
    {
        await using var browserContext = await browser.NewContextAsync();
        var page = await browserContext.NewPageAsync();

        page.GotoAsync("https://localhost/");
        page.ClickAsync("body > header > nav > button");

        await Task.Delay(10000); // This is just for debugging purposes
    }
);
```

**Description**

I have a Blazor Server App that is hosted in IIS. In this App, I have a Button (HTML) that has a Click handler `() => NavigationManager.NavigateTo(url, true);`. All the handler does is navigate to a Login Page.

I wanted to test my Application with Playwright and it works OK when I start only 1-2 instances. When I try with more than 3 instances then it hangs as soon as the click `page.ClickAsync("body > header > nav > button");` is executed. if I close some instances then other instances continue working as they should.

What I have noticed is that if the second parameter in `NavigationManager.NavigateTo(url, true);` is set to `false` then it works.

This can be reproduced with the default Blazor Server Template.

P.S.

If I just run the App with `dotnet run` then it works!

**Comment by mxschmitt** [comment]
I see that you never await these two calls, it should be:

```
        await page.GotoAsync("https://localhost/");
        await page.ClickAsync("body > header > nav > button");
```

Otherwise I don't see something which is wrong, if you could submit a minimal repro, we'd be happy to take a look.

**Comment by dardanlibrary** [comment]
@mxschmitt thanks for the swift reply.

The missing `await` part was my mistake when I was trying to remove the clutter from the code. The actual code has it.

[Here](https://github.com/dardanlibrary/playwright-blazorserver) you can find the minimal repo.

There is a Blazor Server project there [BlazorTests](https://github.com/dardanlibrary/playwright-blazorserver/tree/master/BlazorTests) that has to be published and put on IIS.
There is also a Test project there [BlazorTests.E2ETests](https://github.com/dardanlibrary/playwright-blazorserver/tree/master/BlazorTests.E2ETests) where you can find two tests. One of them passes because it starts with only two Tasks and the other with four Tasks fails.

Let me now if you need something else from me.

**Comment by mxschmitt** [comment]
Oh you are using XUnit, this is discouraged, because their parallelism mode is not compatible with Playwright, see here: https://playwright.dev/dotnet/docs/test-runners#xunit-support

Could you use Nunit or MSTest instead? There it works most likely.

**Comment by dardanlibrary** [comment]
Just tried it with NUnit and I got the same result. I have also updated the repo for you to check.
I have previously checked the link you provided, but the way I understand it, it does not cover my case.
Based on the link, the Test Runner will run tests in parallel but they have to be separate tests.
What I want, is to run the same test many times.

**Comment by mxschmitt** [comment]
I tried to reproduce it, but for me it was working fine. I'd adjust the parallelism model a bit tho, so you don't do CreateAsync all the time. The best would be that you have N workers, e.g. 3 and inside of them you create Playwright instances. These workers stay alive and then work on your jobs over and over again.

This is very similar to what we do in our nunit harness: https://playwright.dev/dotnet/docs/test-runners

**Comment by mxschmitt** [comment]
Sorry for the late reply. I'm unfortunately not familiar with Blazor either but, happy to accept any PRs/research in that direction.

**Comment by dardanlibrary** [comment]
I've been experimenting with this since my last comment. It seems the issue is on the IIS side. I was using IIS on Windows Home and not on Windows Server and based on some online documentation, IIS on Windows Home has a limit on the number of concurrent connections.

I have created a virtual machine with Windows Server and I am not facing the same issue - I was able to execute 100 parallel tests.

Sorry if I bothered you and thanks a lot for your time.
